File: internal/rules/template.go
package rules

import "regexp"

// Message templating for declarative rules.
//
// A fixed set of `{placeholder}` tokens, substituted from what the matcher
// found. Deliberately not an expression language: a rule file decides the
// wording, not the control flow, and the moment a template can compute it
// stops being data.

// templateToken also matches a `${…}` so it can decline to substitute one.
// Rule text is HCL-adjacent — a fix that writes an interpolation would
// otherwise have its inner braces eaten.
var templateToken = regexp.MustCompile(`\$?\{([a-z_]+)\}`)

// expand substitutes known tokens and leaves everything else byte for byte.
//
// Unknown tokens survive untouched on purpose: fix and suggestion templates
// contain literal HCL braces (`variable "x" {`), and a templater that
// swallowed or errored on them could not write Terraform.
func expand(tmpl string, vars map[string]string) string {
	if tmpl == "" {
		return ""
	}
	return templateToken.ReplaceAllStringFunc(tmpl, func(tok string) string {
		if tok[0] == '$' {
			return tok // an interpolation, not a placeholder
		}
		if v, ok := vars[tok[1:len(tok)-1]]; ok {
			return v
		}
		return tok
	})
}

// expandAll applies expand to a slice, for multi-line fix bodies.
func expandAll(tmpls []string, vars map[string]string) []string {
	out := make([]string, len(tmpls))
	for i, t := range tmpls {
		out[i] = expand(t, vars)
	}
	return out
}

File: src/tfpdf/rules/template.py
"""Format declarative rule messages."""

from __future__ import annotations

import re

#: Also matches a `${…}` so it can decline to substitute one. Rule text is
#: HCL-adjacent — a fix that writes an interpolation would otherwise have its
#: inner braces eaten.
_TEMPLATE_TOKEN = re.compile(r"\$?\{([a-z_]+)\}")

_GO_ESCAPES = {
    "\a": r"\a",
    "\b": r"\b",
    "\f": r"\f",
    "\n": r"\n",
    "\r": r"\r",
    "\t": r"\t",
    "\v": r"\v",
    "\\": "\\\\",
    '"': r"\"",
}


def expand(template: str, variables: dict[str, str]) -> str:
    """Replace known template tokens and leave all other text unchanged."""
    if not template:
        return ""

    def repl(m: re.Match[str]) -> str:
        tok = m.group(0)
        if tok[0] == "$":
            return tok  # an interpolation, not a placeholder
        return variables.get(m.group(1), tok)

    return _TEMPLATE_TOKEN.sub(repl, template)


def expand_all(tmpls: list[str], variables: dict[str, str]) -> list[str]:
    """Expand a list of template lines, such as a multiline fix body."""
    return [expand(t, variables) for t in tmpls]


def go_quote(s: str) -> str:
    """Quote a string using Go strconv.Quote conventions."""
    out = ['"']
    for ch in s:
        esc = _GO_ESCAPES.get(ch)
        if esc is not None:
            out.append(esc)
        elif ch.isprintable():
            out.append(ch)
        elif ord(ch) < 0x100:
            out.append(f"\\x{ord(ch):02x}")
        elif ord(ch) < 0x10000:
            out.append(f"\\u{ord(ch):04x}")
        else:
            out.append(f"\\U{ord(ch):08x}")
    out.append('"')
    return "".join(out)
